File: packages/postpin-php/src/Resource/Rates.php
<?php

declare(strict_types=1);

namespace Postpin\Resource;

use Postpin\Http\HttpClient;

/** Shipping-rate calculation. */
final class Rates
{
    public function __construct(private readonly HttpClient $http)
    {
    }

    /**
     * Calculate the shipping rate between two Indian pincodes.
     *
     * Supported $params keys: origin (string), destination (string),
     * weight (int, grams), length/width/height (float, cm),
     * service ("surface"|"express"|"same_day"), cod (bool),
     * declared_value (float, rupees), idempotency_key (string).
     *
     * @param array<string, mixed> $params
     * @return array<string, mixed> the rate result
     */
    public function calculate(array $params): array
    {
        $idempotencyKey = $params['idempotency_key'] ?? null;
        unset($params['idempotency_key']);

        $body = [
            'origin' => $params['origin'] ?? null,
            'destination' => $params['destination'] ?? null,
            'weight' => $params['weight'] ?? null,
        ];
        foreach (['length', 'width', 'height', 'service', 'cod', 'declared_value'] as $key) {
            if (array_key_exists($key, $params)) {
                $body[$key] = $params[$key];
            }
        }

        return (array) $this->http->request('POST', '/rates/calculate', [], $body, $idempotencyKey);
    }
}

File: packages/postpin-python/src/postpin/resources/rates.py
from __future__ import annotations

from typing import Any, Dict, Optional

from .._http import HTTPClient
from ..types import RateResult, ServiceLevel


class Rates:
    def __init__(self, http: HTTPClient) -> None:
        self._http = http

    def calculate(
        self,
        *,
        origin: str,
        destination: str,
        weight: int,
        length: Optional[float] = None,
        width: Optional[float] = None,
        height: Optional[float] = None,
        service: Optional[ServiceLevel] = None,
        cod: Optional[bool] = None,
        declared_value: Optional[float] = None,
        idempotency_key: Optional[str] = None,
        timeout: Optional[float] = None,
    ) -> RateResult:
        """Calculate the shipping rate between two Indian pincodes.

        ``weight`` is in grams; ``length``/``width``/``height`` are in cm and
        enable volumetric-weight pricing. ``declared_value`` is in rupees.
        """
        body: Dict[str, Any] = {"origin": origin, "destination": destination, "weight": weight}
        if length is not None:
            body["length"] = length
        if width is not None:
            body["width"] = width
        if height is not None:
            body["height"] = height
        if service is not None:
            body["service"] = service
        if cod is not None:
            body["cod"] = cod
        if declared_value is not None:
            body["declared_value"] = declared_value

        data, _ = self._http.request(
            "POST",
            "/rates/calculate",
            json_body=body,
            idempotency_key=idempotency_key,
            timeout=timeout,
        )
        return RateResult._from(data)
